<a href="https://colab.research.google.com/github/anulax1114-design/PRAC1-st20347210-DA/blob/main/DA_PRAC_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**INTRODUCTION**
---

**Overview**
-----
Apart from land and water, air is one of the most essential resources for life. However, air pollution remains one of the most serious environmental challenges worldwide. Cities like Beijing have experienced significant air quality issues due to rapid urbanisation, industrial activities, and increasing vehicle emissions. Monitoring and analysing air pollution levels is therefore important to understand its impact on public health and environmental sustainability.

With advancements in technology, large volumes of air quality data are continuously collected from monitoring stations. This data provides valuable insights into pollution levels and helps researchers and policymakers make informed decisions. In this project, programming and data analysis techniques are used to explore, analyse, and model air quality data from Beijing.

**What is Air Quality Index (AQI) and why is it needed?**

Air quality is measured using different pollutants such as PM2.5, PM10, SO₂, NO₂, CO, and O₃. However, interpreting multiple pollutant values can be complex and difficult to understand for the general public.

To simplify this, the Air Quality Index (AQI) was developed. AQI converts complex air pollution data into a single value and category (such as “Good”, “Moderate”, or “Poor”), making it easier for people to understand the quality of air and take necessary precautions.

AQI is important because:

•	It provides a simple way to communicate air pollution level.

•	It helps individuals, especially those with health issues, make informed decisions

•	It supports governments in planning and implementing pollution control measures

**What this Dataset is about**

This dataset is used to analyse and understand air quality patterns in Beijing. It consists of hourly observations collected from multiple monitoring stations between March 2013 and February 2017.

**For this study, four stations were selected:**

•	Urban stations: Dongsi and Guanyuan

•	Suburban stations: Shunyi and Huairou

**Why this datasets were Chosen**

***Dongsi and Guanyuan — Inner Urban Stations***

Dongsi (Chaoyang district) and Guanyuan (Xicheng district) are both classified as urban stations situated inside the fourth ring road of Beijing. Yao et al. (2015) applied cluster analysis to all 12 stations and found that urban stations form distinct clusters with the highest average PM2.5 concentrations, driven by dense residential activity, commercial traffic and road emissions. Xu and Zhang (2020) further confirm that urban areas in south Beijing consistently recorded higher PM2.5 levels than suburban sites at every hour of the day, every day of the week and every month of the year. Selecting stations from two different urban districts — east and west of the city centre — also ensures spatial diversity within the urban zone, avoiding geographic bias.

***Huairou and Shunyi — Outer Suburban Stations***

Huairou (north-west) and Shunyi (north-east) are both officially classified as suburban stations in the outer ring of Beijing. Yao et al. (2015) demonstrate that suburban stations form a separate cluster with meaningfully lower PM2.5 averages, with the outermost stations recording the minimum pollution values across the entire network. Batterman et al. (2016) further report that annual average PM2.5 concentrations varied by up to 20 µg/m³ across the study area, with suburban sites showing consistently lower readings than urban and peri-urban zones — differences associated with higher vegetation cover and lower construction and traffic activity. Selecting Huairou and Shunyi from opposite sides of the suburban ring also provides geographic spread, ensuring the suburban signal is not localised to one direction.

**Why This Combination Is Analytically Strongest**

Yao et al. (2015) demonstrate that the correlation between urban and suburban stations is generally below 0.85, confirming that these two groups capture genuinely different pollution dynamics rather than redundant measurements. This contrast is essential for the EDA tasks in this assessment, where comparing station types is a core objective. The urban–suburban gradient captured by these four stations also directly supports the model building task, as station type becomes a meaningful predictive feature for PM2.5 concentration — consistent with the data-driven approach to pollution monitoring advocated by Sokhi et al. (2022), Xu and Zhang (2020) and Batterman et al. (2016).

The dataset includes:

Air Pollutants

•	PM2.5, PM10

•	SO₂, NO₂, CO, O₃

Meteorological Factors

•	Temperature (Temp)

•	Dew Point (Dewp)

•	Pressure (Pres)

•	Wind Speed (Wspd)

•	Rainfall (Rain)

By analysing this dataset, the project aims to identify pollution patterns, seasonal trends, and differences between urban and suburban areas. Using programming and data analytics tools, the raw data is transformed into meaningful insights that support environmental understanding and decision-making.

**Research Questions**

•	Which pollutants have the greatest influence on PM2.5 levels in Beijing?

•	How do meteorological factors affect air pollution levels?

•	What seasonal and long-term trends can be observed in air quality?

•	Can machine learning models accurately predict PM2.5 levels using historical data?

•	How does model performance vary across different monitoring stations?



### **Import The Necessary Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn import preprocessing
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

**GIT Cloning**

In [2]:
! git config --global user.name "anulax1114-design"
! git config --global user.email "anulax1114@gmail.com"

In [3]:
username = "anulax1114-design"
repo = "PRAC1-st20347210-DA"

In [4]:
! git clone https://@github.com/{username}/{repo}

Cloning into 'PRAC1-st20347210-DA'...
remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 6 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (6/6), 2.78 MiB | 4.74 MiB/s, done.


In [5]:
%cd {repo}

/content/PRAC1-st20347210-DA


In [6]:
%ls

PRSA_Data_Dongsi_20130301-20170228.csv
PRSA_Data_Guanyuan_20130301-20170228.csv
PRSA_Data_Huairou_20130301-20170228.csv
PRSA_Data_Shunyi_20130301-20170228.csv


**Merging of the csv Files**

In [7]:
# Import required libraries
import pandas as pd   # Used for data handling
import glob           # Used to find files
import os             # Used to check directory

# STEP 1: Check current working directory
print(f"Current working directory: {os.getcwd()}")

# STEP 2: Find all Beijing dataset CSV files
csv_files = glob.glob("PRSA_Data_*.csv")

# Create an empty list to store datasets
all_data = []

# STEP 3: Load each dataset
if not csv_files:
    print("No dataset files found. Please check your folder.")
else:
    for file in csv_files:
        # Read CSV file
        df = pd.read_csv(file)

        # Extract station name from file name
        station_name = file.split("_")[2]

        # Add station column
        df["station"] = station_name

        # Append to list
        all_data.append(df)

        print(f"Loaded: {station_name}")

# STEP 4: Merge all datasets into one
if all_data:
    combined_df = pd.concat(all_data, ignore_index=True)

    # Save merged dataset
    combined_df.to_csv("beijing_merged_data.csv", index=False)

    print(f"Successfully merged {len(csv_files)} datasets.")
    print(f"Total rows: {len(combined_df)}")
else:
    print("No data to merge.")

Current working directory: /content/PRAC1-st20347210-DA
Loaded: Shunyi
Loaded: Guanyuan
Loaded: Huairou
Loaded: Dongsi
Successfully merged 4 datasets.
Total rows: 140256


Datasets from the selected Beijing monitoring stations (Dongsi, Guanyuan, Shunyi, and Huairou) were imported using Pandas and combined into a single dataset. A ‘station’ column was added to each dataset prior to merging to retain the source of each observation. The datasets were then concatenated into a unified DataFrame, ensuring a consistent structure for analysis. This approach preserves station-level information, enabling effective comparison of air quality across different locations. The final merged dataset was saved for subsequent preprocessing and analysis.

##**Loading the Pandas Dataframe:**

In [9]:

# Load the merged Beijing dataset into a DataFrame
df = pd.read_csv("beijing_merged_data.csv")
df

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,3.0,6.0,3.0,8.0,300.0,44.0,-0.9,1025.8,-20.5,0.0,NW,9.3,Shunyi
1,2,2013,3,1,1,12.0,12.0,3.0,7.0,300.0,47.0,-1.1,1026.1,-21.3,0.0,NW,9.4,Shunyi
2,3,2013,3,1,2,14.0,14.0,NaN,7.0,200.0,22.0,-1.7,1026.2,-23.0,0.0,NW,8.6,Shunyi
3,4,2013,3,1,3,12.0,12.0,3.0,5.0,NaN,NaN,-2.1,1027.3,-23.3,0.0,NW,6.6,Shunyi
4,5,2013,3,1,4,12.0,12.0,3.0,NaN,200.0,11.0,-2.4,1027.7,-22.9,0.0,NW,4.5,Shunyi
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
140251,35060,2017,2,28,19,16.0,51.0,3.0,29.0,400.0,73.0,12.5,1013.5,-16.2,0.0,NW,2.4,Dongsi
140252,35061,2017,2,28,20,18.0,45.0,3.0,43.0,500.0,54.0,11.6,1013.6,-15.1,0.0,WNW,0.9,Dongsi
140253,35062,2017,2,28,21,23.0,58.0,5.0,61.0,700.0,28.0,10.8,1014.2,-13.3,0.0,NW,1.1,Dongsi
140254,35063,2017,2,28,22,23.0,53.0,9.0,75.0,900.0,15.0,10.5,1014.4,-12.9,0.0,NNW,1.2,Dongsi


**Description of Dataset Columns**

The dataset contains hourly air quality and meteorological measurements collected from multiple monitoring stations in Beijing. The key variables included are:

*  PM2.5 – Fine particulate matter smaller than 2.5 micrometres, which can penetrate deep into the lungs and pose serious health risks.

*  PM10 – Particulate mattersmaller than 10 micrometres, commonly originating from dust and combustion processes.

*  SO₂ (Sulfur Dioxide) – A gas mainly produced from burning fossil fuels, contributing to respiratory problems and acid rain.

* NO₂ (Nitrogen Dioxide) – A harmful gas emitted from vehicles and industrial activities, associated with respiratory issues.

*  CO (Carbon Monoxide) – A colourless, toxic gas formed due to incomplete combustion of fuels.

*  O₃ (Ozone) – A secondary pollutant formed through chemical reactions in sunlight, harmful at ground level.

**Meteorological Variables**


* TEMP (Temperature) - Measures atmospheric temperature, which influences pollutant dispersion.

* DEWP (Dew Point) - Indicates moisture content in the air.

* PRES (Pressure) - Atmospheric pressure, affecting air movement and pollution concentration.

* WSPM (Wind Speed) - Wind speed, which plays a key role in dispersing pollutants.

* RAIN (Rainfall) - Precipitation that can reduce pollution levels by washing out particles.

**Time and Station Information**


* year, month, day, hour - Time-related variables representing when each observation was recorded.


* station - Identifies the monitoring station (Dongsi, Guanyuan, Shunyi, or Huairou), enabling comparison across locations.

##**Exploratory Data Analysis**

In [10]:
df.shape

(140256, 18)

In [11]:
df.columns

Index(['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2',
       'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station'],
      dtype='object')

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140256 entries, 0 to 140255
Data columns (total 18 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   No       140256 non-null  int64  
 1   year     140256 non-null  int64  
 2   month    140256 non-null  int64  
 3   day      140256 non-null  int64  
 4   hour     140256 non-null  int64  
 5   PM2.5    137024 non-null  float64
 6   PM10     137949 non-null  float64
 7   SO2      136843 non-null  float64
 8   NO2      134992 non-null  float64
 9   CO       131706 non-null  float64
 10  O3       135779 non-null  float64
 11  TEMP     140114 non-null  float64
 12  PRES     140112 non-null  float64
 13  DEWP     140109 non-null  float64
 14  RAIN     140110 non-null  float64
 15  wd       139312 non-null  object 
 16  WSPM     140135 non-null  float64
 17  station  140256 non-null  object 
dtypes: float64(11), int64(5), object(2)
memory usage: 19.3+ MB


**Number of Records in Each station**

In [13]:
# Count number of records for each station
stations = df['station'].value_counts()

# Total number of stations
print(f"Total number of stations in the dataset: {len(stations)}")

# Display station-wise counts
stations

Total number of stations in the dataset: 4


,count
station,
Shunyi,35064
Guanyuan,35064
Huairou,35064
Dongsi,35064


In [14]:
df['station'].nunique()

4

In [15]:
df.describe()

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,WSPM
count,140256.000,140256.000000,140256.000000,140256.000000,140256.000000,137024.000000,137949.000000,136843.000000,134992.000000,131706.000000,135779.000000,140114.000000,140112.000000,140109.000000,140110.000000,140135.000000
mean,17532.500,2014.662560,6.522930,15.729637,11.500000,79.579509,102.418485,15.473286,47.076537,1200.947307,57.013350,13.272530,1011.264177,2.568666,0.065119,1.757224
std,10122.141,1.177201,3.448715,8.800123,6.922211,80.433210,91.054730,21.524836,33.320134,1113.648561,56.295033,11.534021,10.443904,13.824228,0.828778,1.246207
min,1.000,2013.000000,1.000000,1.000000,0.000000,2.000000,2.000000,0.285600,1.026500,100.000000,0.214200,-19.900000,982.800000,-43.400000,0.000000,0.000000
25%,8766.750,2014.000000,4.000000,8.000000,5.750000,20.000000,34.000000,2.000000,21.000000,500.000000,12.000000,3.000000,1002.900000,-8.900000,0.000000,0.900000
50%,17532.500,2015.000000,7.000000,16.000000,11.500000,55.000000,80.000000,7.000000,40.000000,900.000000,45.000000,14.300000,1010.900000,3.200000,0.000000,1.400000
75%,26298.250,2016.000000,10.000000,23.000000,17.250000,111.000000,142.000000,19.000000,66.000000,1500.000000,81.000000,23.100000,1019.425000,15.300000,0.000000,2.200000
max,35064.000,2017.000000,12.000000,31.000000,23.000000,941.000000,999.000000,315.000000,270.000000,10000.000000,1071.000000,41.100000,1042.800000,29.100000,72.500000,12.900000


In [16]:
df.isnull().sum()

,0
No,0
year,0
month,0
day,0
hour,0
PM2.5,3232
PM10,2307
SO2,3413
NO2,5264
CO,8550


## **Inference:**


**DATA UNDERSTANDING**

The dataset contains 140,256 records and 18 variables, indicating a large and comprehensive dataset. It includes data from four monitoring stations (Dongsi, Guanyuan, Shunyi, and Huairou), each contributing an equal number of observations, ensuring a balanced dataset for comparison.

Most variables are numerical, representing pollutant concentrations and meteorological factors, while ‘station’ and ‘wd’ are categorical. The statistical summary shows noticeable variation in pollutant levels such as PM2.5 and PM10, indicating fluctuations in air quality over time.

Missing values are present in several pollutant variables (e.g., CO, NO₂, PM2.5), whereas time-related and station columns have no missing values. This highlights the need for preprocessing before further analysis.

Overall, the dataset is well-structured and suitable for analysis, with minor data cleaning required.

### **Datetime Creation**

df: This represents the dataset stored in a DataFrame format.

df[[‘year’,‘month’,‘day’,‘hour’]]: These columns contain the separate components of date and time.

pd.to_datetime(): This function from the pandas library combines these columns and converts them into a single datetime format that Python can easily understand.

So, in summary, this step combines the year, month, day, and hour columns and converts them into a proper datetime format. This is useful because it allows easier time-based analysis such as sorting data, identifying trends, and creating time-series visualisations.

In [17]:
# Create datetime column from year, month, day, hour
df['datetime'] = pd.to_datetime(df[['year','month','day','hour']])


In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140256 entries, 0 to 140255
Data columns (total 19 columns):
 #   Column    Non-Null Count   Dtype         
---  ------    --------------   -----         
 0   No        140256 non-null  int64         
 1   year      140256 non-null  int64         
 2   month     140256 non-null  int64         
 3   day       140256 non-null  int64         
 4   hour      140256 non-null  int64         
 5   PM2.5     137024 non-null  float64       
 6   PM10      137949 non-null  float64       
 7   SO2       136843 non-null  float64       
 8   NO2       134992 non-null  float64       
 9   CO        131706 non-null  float64       
 10  O3        135779 non-null  float64       
 11  TEMP      140114 non-null  float64       
 12  PRES      140112 non-null  float64       
 13  DEWP      140109 non-null  float64       
 14  RAIN      140110 non-null  float64       
 15  wd        139312 non-null  object        
 16  WSPM      140135 non-null  float64    

##**Checking Missing Values**

In [19]:
# Function to evaluate missing data in the dataset
def evaluate_missing_data(data):

    # Count missing entries in each column
    missing_count = data.isnull().sum()

    # Calculate percentage of missing entries
    missing_percentage = (missing_count / len(data)) * 100

    # Combine counts and percentages into one table
    summary_table = pd.concat([missing_count, missing_percentage], axis=1)

    # Rename columns for clarity
    summary_table.columns = ['Total Missing', 'Missing (%)']

    # Sort columns by highest missing percentage
    summary_table = summary_table.sort_values('Missing (%)', ascending=False)

    return summary_table

missing_summary = evaluate_missing_data(df)
display(missing_summary.style.background_gradient(cmap='Reds'))

,Total Missing,Missing (%)
CO,8550,6.095996
NO2,5264,3.753137
O3,4477,3.192020
SO2,3413,2.433407
PM2.5,3232,2.304358
PM10,2307,1.644849
wd,944,0.673055
DEWP,147,0.104808
RAIN,146,0.104095
PRES,144,0.102669


In [20]:
# Missing values for each station
missing_values_per_station = df.groupby('station').apply(lambda x: x.isnull().sum())

missing_values_per_station

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station,datetime
station,,,,,,,,,,,,,,,,,,,
Dongsi,0,0,0,0,0,750,553,663,1601,3197,664,20,20,20,20,78,14,0,0
Guanyuan,0,0,0,0,0,616,429,474,659,1753,1173,20,20,20,20,81,14,0,0
Huairou,0,0,0,0,0,953,777,980,1639,1422,1151,51,53,53,55,302,49,0,0
Shunyi,0,0,0,0,0,913,548,1296,1365,2178,1489,51,51,54,51,483,44,0,0


**Inference (Missing Values Analysis)**

The column-wise analysis shows that missing values are primarily concentrated in pollutant variables. Among these, CO has the highest proportion of missing values (~6.06%), followed by NO₂, O₃, SO₂, and PM2.5, while PM10 also shows a moderate level of missing data.

In contrast, meteorological variables such as temperature, pressure, and rainfall contain very minimal missing values. Additionally, time-related variables and the station column have no missing values, indicating consistency in the data structure.

Overall, the pattern of missing values is consistent across variables, with most data gaps occurring in pollutant measurements. These missing values will need to be addressed during preprocessing to ensure reliable analysis and modelling.

In [ ]:
print("Total missing values per Stations:")
print(df.groupby('station').apply(lambda x: x.isnull().sum().sum()))

**Inference (Missing Values by Station)**

The total number of missing values varies across stations, indicating that data completeness is not entirely uniform.

* Shunyi has the highest number
of missing entries, suggesting comparatively larger gaps in pollutant measurements or data recording.
* Dongsi and Huairou show moderate levels of missing values, indicating some inconsistencies in data collection.
* In contrast, Guanyuan has the lowest number of missing values, making its data relatively more complete and reliable.

Overall, all stations contain missing values, but the differences are not extreme, indicating that data gaps are fairly consistent across locations rather than being heavily concentrated in one station.